jest/verifyComponent.tsx

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+import {render, fireEvent} from '@testing-library/react';
+
+/**
+ * Verify aspects of a component such as ref forwarding, element remapping, and extra prop
+ * spreading. This is a good litmus test for setting up components correctly. Only use on components
+ * that represent an element.
+ * @example
+ * import { MyComponent, useMyComponentModel } from '...'
+ *
+ * describe('MyComponent', () => {
+ *   verifyComponent(MyComponent, { modelHook: useMyComponentModel })
+ *
+ *   // other tests
+ * })
+ */
+export function verifyComponent(
+  Component: React.ComponentType<any>,
+  {modelHook}: {modelHook?: Function}
+) {
+  describe('verifyComponent', () => {
+    const Test = React.forwardRef(({...props}: {as?: React.ReactType}, ref) => {
+      const model = modelHook?.() || null;
+
+      return <Component model={model} ref={ref} {...props} />;
+    });
+
+    it('should forward the ref correctly', () => {
+      const ref = {current: null};
+
+      // test ref - if everything is working, it will be referring to a document element
+      render(<Test ref={ref} />);
+      expect(ref.current).toBeInTheDocument();
+    });
+
+    it('should apply the as prop correctly', () => {
+      // test various `as` props
+      const MyCustomComponent = ({children, ...props}: any) => {
+        return <article {...props}>{children}</article>;
+      };
+
+      const {container, rerender} = render(<Test as="section" />);
+      expect(container).toContainHTML('<section');
+
+      rerender(<Test as="button" />);
+      expect(container).toContainHTML('<button');
+
+      rerender(<Test as={MyCustomComponent} />);
+      expect(container).toContainHTML('<article');
+    });
+
+    it('should forward extra props properly - extra props should overwrite component props', () => {
+      const ref = {current: null};
+      const attrs = {
+        'aria-label': 'MyAriaLabel',
+        id: 'MyID',
+        role: 'MyRole',
+        'data-testid': 'MyTestID',
+      } as const;
+      render(<Test ref={ref} {...attrs} {...callbacks} />);
+      Object.keys(attrs).forEach((key: keyof typeof attrs) => {
+        expect(ref.current).toHaveAttribute(key, attrs[key]);
+      });
+    });
+
+    const callbacks = {
+      onClick: 'click',
+      onKeyDown: 'keyDown',
+      onKeyUp: 'keyUp',
+      onMouseDown: 'mouseDown',
+      onMouseUp: 'mouseUp',
```

Comment:
A limited list of callbacks. `fireEvent` does give me a larger list, but these are the most common ones.

Reply:
I created a separate test for each callback because there is not enough context in the failure to isolate which callback is the problem.